ssl/quic/quic_tls.c

```diff
@@ -702,19 +699,22 @@ int ossl_quic_tls_tick(QUIC_TLS *qtls)
          */
         SSL_set_bio(qtls->args.s, nullbio, nullbio);
 
-        if (qtls->args.is_server) {
+        if (qtls->args.is_server)
             SSL_set_accept_state(qtls->args.s);
-            if (!SSL_set_num_tickets(qtls->args.s, 0)) {
-                qtls->inerror = 1;
-                return 0;
-            }
-        } else {
+        else
             SSL_set_connect_state(qtls->args.s);
-        }
 
         qtls->configured = 1;
     }
-    ret = SSL_do_handshake(qtls->args.s);
+
+    if (qtls->complete)
+        /*
+         * There should never be app data to read, but calling SSL_read() will
+         * ensure any post-handshake messages are processed.
+         */
+        ret = SSL_read(qtls->args.s, NULL, 0);
```

Comment:
That was simpler than I was expecting! :laughing:

Reply:
I assume it would be premature optimization at this point but this seems to call SSL_read() on any tick call. Or am I wrong? I assume there is no point in calling it unless the crypto stream received something from the peer so that could be an optimization.

Reply:
> I assume there is no point in calling it unless the crypto stream received something from the peer so that could be an optimization.

It's not quite as straight forward as that. If multiple handshake messages are sent in a single CRYPTO frame then they can all be passed on in one go and buffered in the TLS code - but SSL_read() will only process one of them at a time. So availability of a new crypto stream data is not necessarily a good indicator by itself of whether we need to call SSL_read(). We might be able to query whether we have anything buffered in order to come up with a sensible check. But I think its premature to add that complexity at the moment.